cpp/common/src/codingstandards/cpp/types/Compatible.qll

```diff
@@ -204,79 +208,204 @@ signature module TypeEquivalenceSig {
+    exists(SpecifiedType t1Spec, SpecifiedType t2Spec |
+      recurses(pragma[only_bind_into](t1Spec), pragma[only_bind_into](t2Spec)) and
+      (
+        t1 = unspecify(t1Spec) and
+        t2 = unspecify(t2Spec)
+      )
+    )
+    or
+    // If function types are compared, their return types and parameter types must be compared.
+    exists(FunctionType t1Func, FunctionType t2Func |
+      recurses(pragma[only_bind_into](t1Func), pragma[only_bind_into](t2Func)) and
+      (
+        t1 = t1Func.getReturnType() and
+        t2 = t2Func.getReturnType()
+        or
+        exists(int i |
+          t1 = t1Func.getParameterType(pragma[only_bind_out](i)) and
+          t2 = t2Func.getParameterType(i)
         )
+      )
+    )
+    or
+    // If the config says to resolve typedefs, and a typedef type is compared to a non-typedef
+    // type, then the non-typedef type must be compared to the base type of the typedef.
+    Config::resolveTypedefs() and
+    exists(TypedefType tdtype |
+      tdtype.getBaseType() = t1 and
+      recurses(pragma[only_bind_into](tdtype), t2)
+      or
+      tdtype.getBaseType() = t2 and
+      recurses(t1, pragma[only_bind_into](tdtype))
+    )
+    or
+    // If two typedef types are compared, then their base types must be compared.
+    exists(TypedefType t1Typedef, TypedefType t2Typedef |
+      recurses(pragma[only_bind_into](t1Typedef), pragma[only_bind_into](t2Typedef)) and
+      (
+        t1 = t1Typedef.getBaseType() and
+        t2 = t2Typedef.getBaseType()
+      )
+    )
   }
 
   /**
-   * A type that is either part of the type subset, or that is reachable from a type in the subset.
+   * Performance related predicate to force top down rather than bottom up evaluation of type
+   * equivalence.
+   *
+   * This predicate is used to determine whether we should recurse on a type. It is used in
+   * conjunction with the `compares` predicate to only recurse on types that are being compared.
+   *
+   * We don't recurse on identical types, as they are already equal. We also don't recurse on
+   * types that are overriden by `Config::overrideTypeComparison`, as that predicate determines
+   * their equalivance.
+   *
+   * For the other types, we have a set of predicates such as `Config::equalPointerTypes` that
+   * holds for `(x, y, true)` if the types `x` and `y` should be considered equivalent when the
+   * pointed-to types of `x` and `y` are equivalent. If the predicate does not hold, or holds for
+   * `(x, y, false)`, then we do not recurse on the types.
+   *
+   * We do not recurse on leaf types.
    */
-  private class RelevantType instanceof Type {
-    RelevantType() { exists(T t | typeGraph*(t, this)) }
-
-    string toString() { result = this.(Type).toString() }
-  }
-
-  private class RelevantDerivedType extends RelevantType instanceof DerivedType {
-    RelevantType getBaseType() { result = this.(DerivedType).getBaseType() }
+  private predicate recurses(Type t1, Type t2) {
```

Comment:
Suggest calling this `shouldRecurseOn` or something similar - that would make the body of the `compares` predicate clearer to read.